06-classification_continued.Rmd

```diff
@@ -559,16 +657,378 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+This is a very difficult problem to solve in
+general, and there are a number of methods that have been developed that apply
+in particular cases of interest; see the additional resources at the end of
+this chapter to find out where you can learn more about variable selection.
+
+The first idea you might think of for a systematic way to select predictors
+is *best subset selection*. Best subset selection essentially involves trying
+all possible subsets of variables and picking the "best" one. In particular, you
+
+1. create a separate model for every possible subset of predictors
+2. tune each one using cross validation
+3. pick the subset of predictors that gives you the highest estimated prediction accuracy  
+
+Best subset selection is applicable to any  classification method (K-NN or otherwise).
+However, it becomes very slow when you have even a moderate
+number of predictors to choose from (say, around 10). This is because the number of possible predictor subsets
+grows very quickly with the number of predictors, and you have to train the model (itself
+a slow process!) for each one. For example, if we have $m=2$ predictors---let's call
+them A and B---then we have 3 variable sets to try: A alone, B alone, and finally A
+and B together. If we have $m=3$ predictors---A, B, and C---then we have 7
+to try: A, B, C, AB, BC, AC, and ABC. In general, the number of models
+we have to train for $m$ predictors is $2^m-1$; in other words, when we 
+get to $m=10$ predictors we have over *one thousand* models to train, and 
+at $m=20$ predictors we have over *one million* models to train! 
+So although it is a simple method, best subset selection is usually too computationally 
+expensive to use in practice.
+
+Another idea is to iteratively build up a model by adding one predictor variable 
+at a time. This method&mdash;known as *forward selection*&mdash;is also widely
+applicable and fairly straightforward. It involves the following steps:
+
+1. start with no predictors
+2. run the following 4 steps until you run out of predictors:
+    1. for each predictor you have not chosen yet, try adding it to the set of predictors to make a candidate model
+    2. tune each candidate model
+    3. pick the model with the highest estimated accuracy
+    4. add the new predictor for that model to your set
+3. pick the set of predictors with a good trade-off between accuracy and model simplicity
+
+Say you have $m$ total predictors to work with. In the first iteration, you have to make
+$m$ candidate models, each with 1 predictor. Then in the second iteration, you have
+to make $m-1$ candidate models, each with 2 predictors (the one you chose before and a new one).
+This pattern continues for as many iterations as you want. If you run the method
+all the way until you run out of predictors to choose, you will end up training
+approximately $m^2$ separate models. This is a *big* improvement from the $2^m-1$
```

Comment:
Should we give some numbers here like we do above and compare? Say p =10? With best subset we said we'd need to train over one thousand models, here with forward stepwise we'd only have to train 100.

Reply:
done